src/main/java/io/apimatic/core/types/pagination/PaginatedData.java

```diff
@@ -0,0 +1,227 @@
+package io.apimatic.core.types.pagination;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+
+import io.apimatic.core.ApiCall;
+import io.apimatic.core.ErrorCase;
+import io.apimatic.core.GlobalConfiguration;
+import io.apimatic.core.HttpRequest;
+import io.apimatic.core.configurations.http.request.EndpointConfiguration;
+import io.apimatic.core.types.CoreApiException;
+import io.apimatic.core.utilities.CoreHelper;
+import io.apimatic.coreinterfaces.http.response.Response;
+
+public class PaginatedData<T, P> implements Iterator<T> {
+
+    private int currentIndex = 0;
+
+    private List<T> data = new ArrayList<T>();
+    private List<P> pages = new ArrayList<P>();
+    private int lastDataSize;
+    private Response lastResponse;
+    private HttpRequest.Builder lastRequestBuilder;
+
+    private TypeReference<P> pageType;
+    private Function<P, List<T>> converter;
+    private PaginationDataManager[] dataManagers;
+    private EndpointConfiguration endpointConfig;
+    private GlobalConfiguration globalConfig;
+
+    /**
+     * @param paginatedData Existing instance to be cloned.
+     */
+    public PaginatedData(final PaginatedData<T, P> paginatedData) {
+        this.pageType = paginatedData.pageType;
```

Comment:
Do we need to copy the `currentIndex` in the copy constructor?

Reply:
no its not required, copy constructor will only copy data but the index will start from 0 always